File: plugins/git-utils/src/commands/pr.ts
// ============================================================
// pr command (← create-pr.sh)
// ============================================================
// CLI: bun run src/cli.ts pr <title> [--description=<d>]
//
// 동작:
//   1. default branch 감지
//   2. 현재 브랜치가 default가 아닌지 확인
//   3. gh auth 확인
//   4. Jira 티켓 감지 → PR 타이틀에 반영
//   5. git push → gh pr create
//
// 기존 create-pr.sh 대비 개선:
//   - push 에러 핸들링 개선 (stderr 무시 제거)
//   - GitHubService 추상화로 테스트 가능
// ============================================================

import type { Command, PrInput, PrOutput } from '../types';
import type { GitService, JiraService, GitHubService } from '../core';

export interface PrDeps {
  git: GitService;
  jira: JiraService;
  github: GitHubService;
}

export type PrCommand = Command<PrInput, PrOutput>;

File: plugins/atelier/cli/src/git/commands/pr.rs
//! `pr` command, ported from git-utils `commands/pr.ts` (originally
//! `create-pr.sh`). Pushes the current branch and opens a PR against the
//! default branch, prefixing the title with a detected Jira ticket.

use crate::git::core::git::GitService;
use crate::git::core::github::GitHubService;
use crate::git::core::jira::detect_ticket;
use crate::git::types::{PrInput, PrOutput};

/// The `pr` command, backed by injected [`GitService`] and [`GitHubService`].
pub struct PrCommand<'a> {
    git: &'a dyn GitService,
    github: &'a dyn GitHubService,
}

impl<'a> PrCommand<'a> {
    pub fn new(git: &'a dyn GitService, github: &'a dyn GitHubService) -> Self {
        Self { git, github }
    }

    pub fn run(&self, input: &PrInput) -> Result<PrOutput, String> {
        if input.title.trim().is_empty() {
            return Err("Title is required".to_string());
        }

        let default_branch = self
            .git
            .detect_default_branch()
            .map_err(|e| e.to_string())?;

        let current = self.git.get_current_branch().map_err(|e| e.to_string())?;
        if current == default_branch {
            return Err(format!(
                "Cannot create PR from default branch ({default_branch})"
            ));
        }

        if !self.github.is_authenticated().map_err(|e| e.to_string())? {
            return Err("GitHub CLI is not authenticated. Run: gh auth login".to_string());
        }

        let ticket = detect_ticket(&current);
        let title = match &ticket {
            Some(t) => format!("[{}] {}", t.normalized, input.title),
            None => input.title.clone(),
        };

        self.git.push(&current, true).map_err(|e| e.to_string())?;

        let body = input.description.clone().unwrap_or_default();
        let url = self
            .github
            .create_pr(&default_branch, &title, &body)
            .map_err(|e| e.to_string())?;

        Ok(PrOutput {
            url,
            title,
            base_branch: default_branch,
            jira_ticket: ticket.map(|t| t.normalized),
        })
    }
}
